components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/InMemoryAPIDeployer.java

```diff
@@ -0,0 +1,159 @@
+                    log.error("Error deploying " + apiId + " in Gateway", e);
+                }
+            } else {
+                log.error("Artifact retriever not found");
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Deploy an API in the gateway using the deployAPI method in gateway admin
+     *
+     * @param assignedGatewayLabels - The labels which the gateway subscribed to
+     * @return True if all API artifacts retrieved from the storage and successfully deployed without any error. else
+     * false
+     */
+    public boolean deployAllAPIsAtGatewayStartup (Set<String> assignedGatewayLabels) {
+
+        if (gatewayArtifactSynchronizerProperties.isRetrieveFromStorageEnabled()) {
+            if (artifactRetriever != null) {
+                try {
+                    for (Iterator<String> it = assignedGatewayLabels.iterator(); it.hasNext();) {
+                        String label = it.next();
+                        List<String> gatewayRuntimeArtifacts = ServiceReferenceHolder
+                                .getInstance().getArtifactRetriever().retrieveAllArtifacts(label);
+                        for (String APIruntimeArtifact :gatewayRuntimeArtifacts){
+                            GatewayAPIDTO gatewayAPIDTO = null;
+                            try {
+                                gatewayAPIDTO = new Gson().fromJson(APIruntimeArtifact, GatewayAPIDTO.class);
+                                log.info("Deploying synapse artifacts of " + gatewayAPIDTO.getName());
+                                apiGatewayAdmin.deployAPI(gatewayAPIDTO);
+                            } catch (AxisFault axisFault) {
+                                log.error("Error in deploying" + gatewayAPIDTO.getName()+ " to the Gateway ");
+                                continue;
+                            }
+                        }
+                    }
+                    return true;
+                } catch (ArtifactSynchronizerException e ) {
+                    log.error("Error  deploying APIs to the Gateway " + e );
+                }
+            } else {
+                log.error("Artifact retriever not found");
+            }
+        }
+        return false;
+    }
+
+    /**
+     * UnDeploy an API in the gateway using the uneployAPI method in gateway admin
+     *
+     * @param apiId        - UUID of the API
+     * @param gatewayLabel - Label of the Gateway
+     * @return True if API artifact retrieved from the storage and successfully undeployed without any error. else false
+     */
+    public boolean unDeployAPI(String apiId, String gatewayLabel) {
+
+        if (gatewayArtifactSynchronizerProperties.isRetrieveFromStorageEnabled() &&
+                gatewayArtifactSynchronizerProperties.getGatewayLabels().contains(gatewayLabel)) {
+            if (artifactRetriever != null) {
+                try {
+                    String gatewayRuntimeArtifact = artifactRetriever
```

Comment:
For undeploy, Do we need to retrieve any info

Reply:
Retrieving the DTO file which has details about removing sequences,localentries etc.

Reply:
ok